default to RDF/XML if no Content-Type header

pmcb55 opened the pull request:
This PR is just a suggestion to address a specific problem encountered with the DOAP vocab served from: http://usefulinc.com/ns/doap#
This vocab is returned without a Content-Type header at all. It just so happens to return RDF/XML, which is why this PR defaults to that serialization format in the absence of a Content-Type value.
Perhaps a better update would be extending the `options` object to `rdfFetch()` include a `contentTypeOverride` value for cases where the developer wants to provide a specific serialization returned by a server who doesn't return a Content Type, but perhaps this DOAP server behaviour is unique enough to not warrant that (i.e., this very old DOAP vocab is the first one that I've encountered (in over 70 public vocabs) that fails to provide a Content-Type header value with it's response), so I think this workaround might be fine for now...?

bergos commented:
@pmcb55 Thanks for the PR!

Vocab hosting is a sad topic. Look at all the [exceptions](https://github.com/zazuko/rdf-vocabularies/blob/master/overrides.ts) required to fetch them.

As this package doesn't come with a formats package, I don't think it's a good idea to set a default value. Also today RDF/XML is probably not a good default anymore.

If you know the content type, I think it would work already to manually set the `content-type` header in the response before you access the quad stream without any code changes in this package.

A more generic solution could be evaluating the `accept` request header and pick the first media type sent in the request. I'm undecided if this is the right way or if it adds too much "magic" to the package. Adding that logic would hide that there was no `content-type` header at all. Maybe it's better to throw very explicit error messages and solve this problem at a higher level. But this package should allow you to implement such logic on a higher level. If setting the `content-type` header in the request doesn't work, we should add a way to allow overwriting it.

pmcb55 commented:
Hi Thomas - yeah, I was being lazy by defaulting to `application/rdf+xml`. So I've updated it to reuse the existing `options.headers.content-type` option, but for the `patchResponse()` function now too (it's only accessible today in the `patchRequest()` function). Updated commits are here: 1e6926e and 51e6364
This update does indeed throw an error if the fetch response has no Content-Type header and no explicit value was provided in the `options` either, as I think that is the best option (rather than trying to heuristically inspect the `Accept:` header).

bergos commented:
@pmcb55 Oh, sorry, I mixed request/response in my last comment. Maybe it's just easier to show some code. I tried it and it works already out of the box with the current version:

```javascript
const fetch = require('../fetch-lite')
const formats = require('../formats-common')

const url = 'http://usefulinc.com/ns/doap#'

async function main () {
  const res = await fetch(url, { formats: formats })

  res.headers.set('content-type', 'application/rdf+xml')

  const stream = await res.quadStream()

  stream.on('data', quad => {
    console.log(quad)
  })
}

main()
```

pmcb55 commented:
Ah ha - brilliant :) !

I had tried to do that myself, but I was trying it in the wrong place - i.e., I had attempted:
```javascript
      return rdfFetch(inputResource, {
        factory: rdf,
        headers: {
          accept: acceptHeader,
          'content-type': 'application/rdf+xml'  // DOESN'T MAKE SENSE HERE - SILLY ME!
        },
        formats,
      })
        .then((rdfResponse) => {
          // THIS IS WHERE I NEEDED TO EXPLICITLY SET THE CONTENT-TYPE!
          //   rdfResponse.headers.set('content-type', 'application/rdf+xml');
          return rdfResponse.dataset();
        })
        .catch((error) => {
          throw new Error('...');
        });
```
So I'm all good again now - thanks so much!

I do still think throwing an exception in function `attachQuadStream()` makes sense though if at that stage there is no Content-Type header value set (otherwise it'll continue blowing up with the unhelpful error `Cannot read property 'split' of null` from [this line](https://github.com/rdfjs-base/fetch-lite/blob/e7121e53d83b46cfc467972866738845f2dbc97d/lib/attachQuadStream.js#L6). And also adding a 'contentTypeFallback' option to the original `fetch()` call probably makes sense too - rather than having to explicitly handle this edge-case in the developers code like I have to do above (the option would still force the developer to add that option and it's value to the `fetch()` call options, but at least it could then be documented and it's all handled in one place rather than two places).

If you agree, I'd be happy enough to re-jig my PR to offer those changes (or I'd probably start again with maybe two fresh more targeted PRs, i.e., one to add the ContentTypeFallback option, and one to throw if no content-type (and no fallback)) - otherwise, feel free to close out this PR (since the above approach does work too).

bergos commented:
Let's keep `res.headers.set()` as the only way to handle this edge case. It's based on the fetch standard interfaces and fit's with the idea of the library to add only the bare minimum required to handle RDF data.

But you are right, it should be documented. Also, the absence of the `content-type` header should give a more readable error message. For both topics, PRs are very welcome.

pmcb55 commented:
Yeah, I agree - closing this PR, and created this one instead: https://github.com/rdfjs-base/fetch-lite/pull/10
